### ISSUE TITLE
stale-while-revalidate in AMP Cache

### ISSUE THREAD

**Issue by choumx** [ampproject/amphtml#21509]
## Background

- [2010] [RFC](https://tools.ietf.org/html/rfc5861#section-3)
- [2014] [Blogpost from RFC author](https://www.mnot.net/blog/2014/06/01/chrome_and_stale-while-revalidate)
- [2014] [Chromium-dev thread and suggested SwR values](https://groups.google.com/a/chromium.org/forum/#!msg/chromium-dev/zchogDvIYrY/ZqWSdt3LJdMJ)
- [2014] [Chrome PRD](https://docs.google.com/document/d/1izgyCOcRtz5Kc0rFtZ68p5mj3E__Uvgpf-5cJJHi7jE/edit#heading=h.mpx9zmbphg16)
- [2018] Chrome [design doc](https://docs.google.com/document/d/1wqbhCdf3eCv-VuUoV34xD2KfMinOCxH4fRe5EDCtXIo/edit#heading=h.st1wclnh8tti), [Intent to Implement](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/P8bZxeA84Z4/gGWF1_4JBAAJ), and [Intent to Ship](https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/rspPrQHfFkI/rrhGyO4DBwAJ)
- [Chrome Platform Status](https://www.chromestatus.com/feature/5050913014153216)

By design, SwR trades off:
1. End-user perceived latency
2. Freshness of responses
3. Traffic to AMP Cache

This issue proposes to improve (1) at some cost to (2) and no impact to (3).

## Status quo

| file       | max-age     | stale-while-revalidate | Σ       |
| ---------- | ----------- | ---------------------- | ------- |
| v0.js      | 3000 (50m)  | 2592000 (30d)          | 30d 50m |
| extensions | 604800 (7d) | 2592000 (30d)          | 37d     |

- 50m `max-age` for v0.js allows fast delivery of cherry-pick releases to users.
- 7d `max-age` for extensions aligns with AMP release cycle. OK for bug fixes due to version-locking feature in v0.js.
- Today, `stale-while-revalidate` is effectively 0 since no browser implements it yet.
- We only care about unversioned URLs since versioned URLs have immutable content.

## Considerations

Let `R` = AMP release interval (currently 7d).

1. `Σ < R` is not valuable since it's too short to provide background revalidation of a new AMP release.

2. `Σ > R` means publishers using latest AMP features risk breaking some users for one pageview (if publisher pages are fresh and AMP is stale). 
    - May be mitigated with publisher messaging, e.g. wait 7d before deploying new hotness, or use SwR of ~7d for publisher pages.

3. `SwR > 0` means magnified impact of AMP JS bugs because # of impacted users increases. 
    - Before: "users with cached JS for up to 50m after fix release".
    - After: Above, plus "users with stale JS for up to `Σ` after culprit release (for one pageview)".
    - May be mitigated by user habit of hitting "refresh" when things go wrong, or by having fewer bugs. :)

## Proposal

| file       | max-age     | stale-while-revalidate | Σ   |
| ---------- | ----------- | ---------------------- | --- |
| v0.js      | 3000 (50m)  | 1206600 (13d 23h 10m)  | 14d |
| extensions | 604800 (7d) | 604800 (7d)            | 14d |

- Improves user-perceived latency for 7DAU e.g. Sunday-only visitors (1).
- Caps the staleness of users' cached AMP to one AMP release (2).
- Limits the impact magnification of AMP JS bugs to 7DAU (3).

## Future work

- If we can handle the extra revalidation traffic, we can reduce `max-age` (and increase SwR) to reduce 50m cache window for bug fixes. TODO: Can we only do this for Chrome UAs?
- Change `max-age` and SwR from fixed values to vary depending on "time to next release". This makes caching/revalidation efficacy independent of time of request.

## Questions

> Can we measure SwR efficacy (e.g. number of background revalidations per day) on AMP Cache?

TBD.

> How does this interact with AMP service worker? 

[No SW visibility in current Chrome implementation](https://groups.google.com/a/chromium.org/d/msg/blink-dev/rspPrQHfFkI/gRu6mPzFDwAJ).

> How are diversions impacted?

- Today (weekly releases), same SwR as non-diverted is fine. 
- With daily releases, a small % of users that have SwR window overlapping two weekly releases may be missing an extra week of AMP functionality for one pageview. Fixable by having variable SwR but not sure it's worthwhile.

/cc @cramforce

**Comment by cramforce** [comment]
I like the plan. I don't think we can make changes to `max-age` only for browsers with swr support.

A more conservative approach would be to use a swr time for `v0.js` of 7d and remove swr for extensions.

**Comment by choumx** [comment]
That can work but the likelihood of revalidation would highly depend on when you visit an AMP page (with clean cache).

- Visit on Tuesday after a release, revalidation window for next release is O(hours).
- Visit on Monday before a release, revalidation window is ~6 days.

I think we should minimum do 8d so daily active users are guaranteed to receive the benefit. Can increase from there.

Ideally, SwR (and max-age) wouldn't be static but vary depending on time to next release. Added that to "future work".

**Comment by cramforce** [comment]
Daily actives would never have a cache miss under both scenarios, right?
Since max-age is 1h, they get a daily revalidation and thus a daily restart
of the N day SWR window.

I don't have a strong opinion on the right choice, just wanted to put an
alternative with slightly different tradeoffs out there.

**Comment by choumx** [comment]
Ah, 304 Not Modified resets freshness. I guess that makes sense.

Ok, will wait a bit to allow discussion. Thanks for the feedback.

**Comment by jridgewell** [comment]
One area that's not considered is the increased extensions max-age compared to v0.js. It's true that version locking will save it, but there's still an expense to parse the response (or keep the parsed response from cache), wait till v0.js initializes, check version lock, _then finally_ issue a request to a rtv'd extension URL to guarantee the version.

As SWR support increases, we should probably lower the max-age for extensions and increase the SWR keep the 14 day window.

**Comment by choumx** [comment]
SWR is being [enabled in Chrome 75](https://www.chromestatus.com/feature/5050913014153216) which becomes [stable on June 4](https://www.chromestatus.com/features/schedule).

**Comment by twifkak** [comment]
Is there is a real concern that usage of new features will break a significant population? Do we have an estimate of what fraction of pageviews is the first one after a new release? If it's significant, then I suspect publisher messaging is insufficient. (What is the medium for that message? How many people will receive it? How will they change their publication processes to prevent such a breakage?)

Brainstorming some ideas:
 - Make a policy of validator additions trailing runtime additions by one week.
 - Prerelease API-compatible runtime stubs by one week that allow the runtime to degrade gracefully.

cc @ampproject/wg-caching

**Comment by choumx** [comment]
> Is there is a real concern that usage of new features will break a significant population?

IMO, this is not a major concern. This will only affect new AMP pages that rely on feature X that are published within ~1 week of feature X's release. And even then, the impact is only that feature X is missing, and for a single page view per user. 

"Feature X is missing" will likely be mostly transparent to users since features are usually incremental in nature. E.g. this won't affect new extensions.

My messaging suggestion was that publishers who _are_ tracking our feature releases that closely typically are also communicating with us on GitHub or Slack, so we could give those pubs a heads up.

**Comment by choumx** [comment]
The change has been submitted for AMP Cache and will be rolled out shortly.
